v2/src/log_sink_etw.c

```diff
@@ -0,0 +1,247 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#include <stdio.h>
+#include <stdarg.h>
+
+#include "windows.h"
+
+#include "TraceLoggingProvider.h"
+#include "evntrace.h"
+
+#include "macro_utils/macro_utils.h"
+
+#include "c_logging/logger.h"
+#include "c_logging/log_sink_etw.h"
+
+TRACELOGGING_DEFINE_PROVIDER(
+    g_my_component_provider,
+    "block_storage_2",
+    (0xDAD29F36, 0x0A48, 0x4DEF, 0x9D, 0x50, 0x8E, 0xF9, 0x03, 0x6B, 0x92, 0xB4));
+/*DAD29F36-0A48-4DEF-9D50-8EF9036B92B4*/
+
+static volatile LONG isETWLoggerRegistered = 0;
+
+static void lazyRegisterEventProvider(void)
+{
+    /*lazily init the logger*/
+    LONG state;
+    while ((state = InterlockedCompareExchange(&isETWLoggerRegistered, 1, 0)) != 2) /*0 - not init, 1 - initializing, 2 - initialized*/
+    {
+        if (state == 0)
+        {
+            /* register the provider*/
+            TLG_STATUS t = TraceLoggingRegister(g_my_component_provider);
+            if (SUCCEEDED(t))
+            {
+                (void)InterlockedExchange(&isETWLoggerRegistered, 2);
+                printf("block_storage_2 ETW provider was registered succesfully (self test). Executable file full path name = %s", _pgmptr); /*_pgmptr comes from https://docs.microsoft.com/en-us/windows/win32/api/libloaderapi/nf-libloaderapi-getmodulefilenamea */
+            }
+            else
+            {
+                (void)printf("block_storage_2 ETW provider was NOT registered.");
+                (void)InterlockedExchange(&isETWLoggerRegistered, 0);
+            }
+        }
+    }
+}
+
+__pragma(pack(push, 1))
+typedef struct SELF_DESCRIBED_EVENT_TAG
+{
+    uint8_t _tlgBlobTyp;
+    uint8_t _tlgChannel;
+    uint8_t _tlgLevel;
+    uint8_t _tlgOpcode;
+    uint64_t _tlgKeyword;
+    uint16_t _tlgEvtMetaSize;
+    uint16_t _tlgEvtTag;
+    uint8_t metadata[];
+} SELF_DESCRIBED_EVENT;
+__pragma(pack(pop))
+
+#define MAX_SELF_DESCRIBED_EVENT_WITH_METADATA (sizeof(SELF_DESCRIBED_EVENT) + 4096)
+
+// This function is done with a little bit of reverse engineering of TraceLogging and guidance from 
+// the TraceLogging.h header about the format of the self described events
+static void emit_self_described_event(const char* event_name, uint32_t trace_level, LOG_CONTEXT_PROPERTY_VALUE_PAIR* context_property_value_pairs, uint32_t property_value_count, const char* message, const char* file, const char* func, int line)
```

Comment:
>context_property_value_pairs [](http://example.com/codeflow?start=117&length=28)

It looks like the structure of these properties is "flat", there's no hierarchy in the properties. Like "parent properties" /"grandparent /self.

Is that somethign tha tETW would support? Can ETW emit self-described structures?

It is awesome as is, and a great improvement1

Reply:
ETW can emit actually self described structures. It was also lurking in my mind how I could make use of that to group the properties of a given context. I don't think it should be that hard, so challenge accepted!

Reply:
A POC of using structs in ETW has been done. perfview shows those decent, Geneva does not actually show the struct info :-). Maybe we can ask for some support for that at some point.